src/server/game/AI/CoreAI/PetAI.cpp

```diff
@@ -89,9 +89,43 @@ void PetAI::_doMeleeAttack()
 
 bool PetAI::_canMeleeAttack() const
 {
-    return me->GetEntry() != 416 /*ENTRY_IMP*/ &&
-        me->GetEntry() != 510 /*ENTRY_WATER_ELEMENTAL*/ &&
-        me->GetEntry() != 37994 /*ENTRY_WATER_ELEMENTAL_PERM*/; 
+    bool canAttack = true;
+    switch (me->GetEntry())
+    {
+        case ENTRY_IMP:
+        case ENTRY_WATER_ELEMENTAL:
+        case ENTRY_WATER_ELEMENTAL_PERM:
+        {
+            canAttack = false;
+            for (uint8 i = 0; i < me->GetPetAutoSpellSize(); ++i)
+            {
+                uint32 spellID = me->GetPetAutoSpellOnPos(i);
+                switch (spellID)
+                {
+                    case IMP_FIREBOLT_RANK_1:
+                    case IMP_FIREBOLT_RANK_2:
+                    case IMP_FIREBOLT_RANK_3:
+                    case IMP_FIREBOLT_RANK_4:
+                    case IMP_FIREBOLT_RANK_5:
+                    case IMP_FIREBOLT_RANK_6:
+                    case IMP_FIREBOLT_RANK_7:
+                    case IMP_FIREBOLT_RANK_8:
+                    case IMP_FIREBOLT_RANK_9:
+                    case WATER_ELEMENTAL_WATERBOLT_1:
+                    case WATER_ELEMENTAL_WATERBOLT_2:
+                        canAttack = true;
+                        break;
```

Comment:
why don't we just `return true;` here?

Reply:
Because you don't need to change it from true to false.

Reply:
When creatures equal entries IMP or water elemental default change to false and when equal true auto spell size pet>0
When again pet want to attack , this again check and if auto spells change to disable (default is false) when check spells, auto spell = 0 and can't attack.

Reply:
I'm sorry my English is poor.

Reply:
yes, but inside the `for` loop I don't see any other assignment to `canAttack` other than `true` on line 116

Reply:
You mean to delete the variable?

Reply:
You mean to delete variable?

Reply:
@Nightprince yes, and replace line 116 with `return true;` and line 128 with `return false;`.

It should work, shouldn't it? and it would be more efficient

Reply:
> @Nightprince yes, and replace line 116 with `return true;` and line 128 with `return false;`.
> 
> It should work, shouldn't it? and it would be more efficient

finish.

Reply:
> @Nightprince yes, and replace line 116 with `return true;` and line 128 with `return false;`.
> 
> It should work, shouldn't it? and it would be more efficient

Thank you for your accuracy